### PR TITLE
kv/RocksDBStore: don't use real wholespace iterator for prefixed access

### PR DESCRIPTION
**Pull request by ifed01**
Signed-off-by: Igor Fedotov <igor.fedotov@croit.io>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by cfsnyder — APPROVED** [review]
lgtm!

File: src/kv/RocksDBStore.cc [hunk 1]
```diff
@@ -3012,7 +3011,13 @@ KeyValueDB::Iterator RocksDBStore::get_iterator(const std::string& prefix, Itera
         std::move(bounds));
     }
   } else {
-    return KeyValueDB::get_iterator(prefix, opts);
+    // use wholespace engine if no cfs are configured
+    // or use default cf otherwise as there is no
+    // matching cf for the specified prefix.
+    auto w_it = cf_handles.size() == 0 || prefix.empty() ?
```
Comment (aclamk) [inline src/kv/RocksDBStore.cc:3017]: `prefix` cannot be empty for Iterator interface.
There is no way to inspect prefix associated with the key.
I guess that it is an assert condition.

File: src/kv/RocksDBStore.cc [hunk 2]
```diff
@@ -667,11 +667,10 @@ rocksdb::ColumnFamilyHandle *RocksDBStore::get_cf_handle(const std::string& pref
  * CF handle. In all other cases, we return a nullptr to indicate that the specified bounds cannot necessarily be mapped
  * to a single CF.
  */
-rocksdb::ColumnFamilyHandle *RocksDBStore::get_cf_handle(const std::string& prefix, const IteratorBounds& bounds) {
```
Comment (aclamk) [inline src/kv/RocksDBStore.cc:670]: Nice refactor - cleanup.

File: src/kv/KeyValueDB.h [hunk 3]
```diff
@@ -322,6 +322,12 @@ class KeyValueDB {
       return generic_iter->status();
     }
   };
+protected:
```
Comment (aclamk) [inline src/kv/KeyValueDB.h:325]: #A

**Review by aclamk — APPROVED** [review]
Good work.
One point for improvement found.

File: src/kv/RocksDBStore.cc (hunk 1 above)
Comment (ifed01) [reply src/kv/RocksDBStore.cc:3017]: will fix in a follow-up PR as the current one is under QA atm and the issue is rather minor IMO

**Comment by ljflores** [comment]
Rados suite review: http://pulpito.front.sepia.ceph.com/?branch=wip-yuri11-testing-2023-03-08-1220

Failures, unrelated:
    1. https://tracker.ceph.com/issues/58585
    2. https://tracker.ceph.com/issues/58560
    3. https://tracker.ceph.com/issues/58946
    4. https://tracker.ceph.com/issues/49287
    5. https://tracker.ceph.com/issues/57755
    6. https://tracker.ceph.com/issues/52316
    7. https://tracker.ceph.com/issues/58496

Details:
    1. rook: failed to pull kubelet image - Ceph - Orchestrator
    2. test_envlibrados_for_rocksdb.sh failed to subscribe to repo - Infrastructure
    3. cephadm: KeyError: 'osdspec_affinity' - Ceph - Orchestrator
    4. SELinux Denials during cephadm/workunits/test_cephadm - Ceph - Orchestrator
    5. task/test_orch_cli: test_cephfs_mirror times out - Ceph - Orchestrator
    6. qa/tasks/mon_thrash.py: _do_thrash AssertionError len(s['quorum']) == len(mons) - Ceph - RADOS
    7. osd/PeeringState: FAILED ceph_assert(!acting_recovery_backfill.empty()) - Ceph - RADOS

**Pull request merged**